File: src/web/frontend/robot_apps_flutter/packages/rmf_api/lib/src/models/building_map.dart
class BuildingMap {
  BuildingMap.fromJson(Map<String, dynamic> json)
      : name = json['name'] as String,
        levels = (json['levels'] as List<dynamic>).map((e) => Level.fromJson(e as Map<String, dynamic>)).toList();

  final String name;
  final List<Level> levels;
}

class Level {
  Level.fromJson(Map<String, dynamic> json)
      : name = json['name'] as String,
        // ignore: avoid_dynamic_calls
        vertices = (json['nav_graphs'][0]['vertices'] as List<dynamic>).map((e) => Vertice.fromJson(e as Map<String, dynamic>)).toList();

  final String name;
  final List<Vertice> vertices;
}

class Vertice {
  Vertice.fromJson(Map<String, dynamic> json)
      : x = json['x'] as double,
        y = json['y'] as double,
        name = json['name'] as String {
    final params = _readParams(json['params'] as List<dynamic>);
    isDispenser = params['pickup_dispenser'] ?? false;
    isIngestor = params['dropoff_ingestor'] ?? false;
  }

  final double x;
  final double y;
  final String name;
  late bool isDispenser;
  late bool isIngestor;

  static Map<String, bool> _readParams(List<dynamic> params) {
    final verticesParams = <String, bool>{};
    for (final param in params) {
      verticesParams[(param as Map<String, dynamic>)['name'] as String] = true;
    }
    return verticesParams;
  }
}

class VerticeParam {
  VerticeParam.fromJson(Map<String, dynamic> json) : name = json['name'] as String;

  final String name;
}

File: src/web/middleware/models/building_map.py
from dataclasses import dataclass
from typing import List, Dict, Any


@dataclass
class Vertice:
    x: float
    y: float
    name: str
    is_dispenser: bool = False
    is_ingestor: bool = False

    @classmethod
    def from_json(cls, json: Dict[str, Any]) -> "Vertice":
        params = cls._read_params(json.get("params", []))
        return cls(
            x=json.get("x", 0.0),
            y=json.get("y", 0.0),
            name=json.get("name", ""),
            is_dispenser=params.get("pickup_dispenser", False),
            is_ingestor=params.get("dropoff_ingestor", False),
        )

    @staticmethod
    def _read_params(params: List[Dict[str, Any]]) -> Dict[str, bool]:
        vertices_params = {}
        for param in params:
            vertices_params[param.get("name", "")] = True
        return vertices_params


@dataclass
class Level:
    name: str
    vertices: List[Vertice]

    @classmethod
    def from_json(cls, json: Dict[str, Any]) -> "Level":
        return cls(
            name=json.get("name", ""),
            vertices=[
                Vertice.from_json(v)
                for v in json.get("nav_graphs", [{}])[0].get("vertices", [])
            ],
        )


@dataclass
class BuildingMap:
    name: str
    levels: List[Level]

    @classmethod
    def from_json(cls, json: Dict[str, Any]) -> "BuildingMap":
        return cls(
            name=json.get("name", ""),
            levels=[Level.from_json(l) for l in json.get("levels", [])],
        )
